Localstorage exception 20176 (#20673)
mdn/content: files/en-us/web/api/window/localstorage/index.md

### Exceptions

- `SecurityError`

>   - : The request violates a policy decision, or the origin is not [a valid scheme/host/port tuple](/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin) (this can happen if the origin uses the `file:` or `data:` schemes, for example). For example, the user may have their browser configured to deny permission to persist data for the specified origin.
> 
>     Note that this can happen if the user has blocked cookies.

Reviewer: ```suggestion
  - : Thrown in one of the following cases:
    - The request violates a policy decision. For example, the user has their browser configured to deny permission to persist data for the specified origin, such as when cookies are blocked.
    - The origin is not [a valid scheme/host/port tuple](/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin). For example, if the origin uses the `file:` or `data:` schemes.
```

(I assume "blocked cookies" is a subset of "deny permission to persist data for the specified origin")

Reviewer 2: This is OK, but I like the current wording better because I think it makes it harder to miss the connection between blocking cookies and localStorage not working, which is an important and (to many) surprising thing.

I think people are fairly often going to come across `localStorage` not working because of this, and will visit this page to see what's up.

In the current wording this fact gets its own paragraph at the end of the section:

<img width="824" alt="Screen Shot 2022-09-19 at 4 50 31 PM" src="https://user-images.githubusercontent.com/432915/191138493-8296126c-c9fb-4d66-88ba-cffd98581c2c.png">

...which I think makes it a lot easier to see than in the suggestion, which is a 4 word phrase at the end of the penultimate bullet:

<img width="823" alt="Screen Shot 2022-09-19 at 4 52 58 PM" src="https://user-images.githubusercontent.com/432915/191138684-b7753d2a-19bd-4c8f-8428-46296303ecce.png">

That was my thinking in giving this point its own paragraph.

Author: I tend to agree with @wbamberg here. It may not be immediately obvious that blocking cookies would block this functionality. Separating it out makes this fact more obvious to the reader.

Reviewer: I agree with the rationale but I strongly object the current organization which is not logical at all. Maybe we can find a way to both highlight the cookies case and the two separate invariants.

Reviewer 2: OK, how about:

```suggestion
  - : Thrown in one of the following cases:

    - The origin is not [a valid scheme/host/port tuple](/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin). This can happen if the origin uses the `file:` or `data:` schemes, for example.
    - The request violates a policy decision. For example, the user has configured the browsers to prevent the page from persisting data.

    Note that if the user blocks cookies, browsers will probably interpret this as an instruction to prevent the page from persisting data.
```

This separates the two reasons, highlights the cookie case, but makes it clear (I hope) that the cookie case is a case of "policy decision".

Reviewer: Looks great to me 👍

- `SecurityError`

>   - : The request violates a policy decision, or the origin is not [a valid scheme/host/port tuple](/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin) (this can happen if the origin uses the `file:` or `data:` schemes, for example). For example, the user may have their browser configured to deny permission to persist data for the specified origin.
> 
> - `Uncaught DOMException`
>   - : The request is blocked by user's browser privacy settings. In some browsers, blocking cookies results in disabling many APIs that persist user session data such as localStorage and [sessionStorage](/en-us/web/api/window/sessionStorage).
> 

Reviewer 2: ```suggestion
- `SecurityError`

  - : The request violates a policy decision, or the origin is not [a valid scheme/host/port tuple](/en-US/docs/Web/Security/Same-origin_policy#definition_of_an_origin) (this can happen if the origin uses the `file:` or `data:` schemes, for example). For example, the user may have their browser configured to deny permission to persist data for the specified origin.

    Note that this can happen if the user has blocked cookies.

```

Author: Ah, this wasn't immediately obvious to me, but after taking another look I agree. Thanks for the feedback! I've committed the changes for review.
